autoarray/structures/grids.py

```diff
@@ -1497,54 +1374,337 @@ def squared_distances_from_coordinate(self, coordinate=(0.0, 0.0)):
+    pixelization_grid : ndarray
+        The (y,x) centre of every Voronoi pixel in scaleds.
+    origin : (float, float)
+        The scaled origin of the Voronoi pixelization's coordinate system.
+    pixel_neighbors : ndarray
+        An array of length (voronoi_pixels) which provides the index of all neighbors of every pixel in \
+        the Voronoi grid (entries of -1 correspond to no neighbor).
+    pixel_neighbors_size : ndarrayy
+        An array of length (voronoi_pixels) which gives the number of neighbors of every pixel in the \
+        Voronoi grid.
+    """
+
+    def __new__(
+        cls, grid, nearest_pixelization_1d_index_for_mask_1d_index=None, *args, **kwargs
+    ):
+        """A pixelization-grid of (y,x) coordinates which are used to form the pixel centres of adaptive pixelizations in the \
+        *pixelizations* module.
+
+        A *PixGrid* is ordered such pixels begin from the top-row of the mask and go rightwards and then \
+        downwards. Therefore, it is a ndarray of shape [total_pix_pixels, 2]. The first element of the ndarray \
+        thus corresponds to the pixelization pixel index and second element the y or x arc -econd coordinates. For example:
+
+        - pix_grid[3,0] = the 4th unmasked pixel's y-coordinate.
+        - pix_grid[6,1] = the 7th unmasked pixel's x-coordinate.
+
+        Parameters
+        -----------
+        pix_grid : ndarray
+            The grid of (y,x) scaled coordinates of every image-plane pixelization grid used for adaptive source \
+            -plane pixelizations.
+        nearest_pixelization_1d_index_for_mask_1d_index : ndarray
+            A 1D array that maps every grid pixel to its nearest pixelization-grid pixel.
+        """
+
+        if isinstance(grid, list):
+            grid = np.asarray(grid)
+
+        obj = grid.view(cls)
+        obj.nearest_pixelization_1d_index_for_mask_1d_index = (
+            nearest_pixelization_1d_index_for_mask_1d_index
+        )
+        obj.interpolator = None
+
+        obj.pixels = grid.shape[0]
+
+        try:
+            obj.voronoi = scipy.spatial.Voronoi(
+                np.asarray([grid[:, 1], grid[:, 0]]).T, qhull_options="Qbb Qc Qx Qm"
+            )
+        except ValueError or OverflowError or scipy.spatial.qhull.QhullError:
+            raise exc.PixelizationException()
+
+        pixel_neighbors, pixel_neighbors_size = pixelization_util.voronoi_neighbors_from_pixels_and_ridge_points(
+            pixels=obj.pixels, ridge_points=np.asarray(obj.voronoi.ridge_points)
+        )
+
+        obj.pixel_neighbors = pixel_neighbors.astype("int")
+        obj.pixel_neighbors_size = pixel_neighbors_size.astype("int")
+        obj.nearest_pixelization_1d_index_for_mask_1d_index = (
+            nearest_pixelization_1d_index_for_mask_1d_index
+        )
+
+        return obj
+
+    def __array_finalize__(self, obj):
+
+        if hasattr(obj, "pixel_neighbors"):
```

Comment:
you could do this as a loop over a tuple of strings with try setattr getattr except AttributeError